From 0cfd2c7b0d54c37c562c6993ec0aaa0971e7da41 Mon Sep 17 00:00:00 2001
From: Nick Hainke <vincent@systemli.org>
Date: Sat, 3 Oct 2026 20:38:44 +0200
Subject: [PATCH 1/3] ujail: cgroups: add cgroups_open_current() helper

Add a helper which opens the cgroup v2 directory the calling process
lives in, as listed in the "0::<path>" line of /proc/self/cgroup. It
complements cgroups_open_dir(), which opens the cgroup of an OCI
container.

The VRF support added in the next commit uses it to create its cgroup
below the cgroup procd started the instance in.

Assisted-by: Claude:claude-opus-5-5
Signed-off-by: Nick Hainke <vincent@systemli.org>
---
 jail/cgroups.c | 32 ++++++++++++++++++++++++++++++++
 jail/cgroups.h |  1 +
 2 files changed, 33 insertions(+)

diff --git a/jail/cgroups.c b/jail/cgroups.c
index 10e4b3f..112a372 100644
--- a/jail/cgroups.c
+++ b/jail/cgroups.c
@@ -65,6 +65,38 @@ void cgroups_prepare(void) {
 	destroyed = false;
 }
 
+/*
+ * Open the cgroup v2 directory the calling process lives in, as listed in
+ * the "0::<path>" line of /proc/self/cgroup.
+ */
+int cgroups_open_current(void)
+{
+	char line[PATH_MAX], path[PATH_MAX];
+	char *nl;
+	FILE *f;
+	int fd = -1;
+
+	f = fopen("/proc/self/cgroup", "re");
+	if (!f)
+		return -1;
+
+	while (fgets(line, sizeof(line), f)) {
+		if (strncmp(line, "0::/", 4))
+			continue;
+
+		nl = strchr(line, '\n');
+		if (nl)
+			*nl = '\0';
+
+		snprintf(path, sizeof(path), "%s%s", CGROUP_ROOT, line + 4);
+		fd = open(path, O_PATH | O_DIRECTORY | O_CLOEXEC);
+		break;
+	}
+
+	fclose(f);
+	return fd;
+}
+
 void cgroups_init(const char *p) {
 	avl_init(&cgvals, avl_strcmp, false, NULL);
 	cgroup_path = strdup(p);
diff --git a/jail/cgroups.h b/jail/cgroups.h
index e6cd464..d6f5d31 100644
--- a/jail/cgroups.h
+++ b/jail/cgroups.h
@@ -33,6 +33,7 @@ int cgroups_open_attr(const char *attr);
 int cgroups_open_dir(void);
 void cgroups_free(void);
 void cgroups_prepare(void);
+int cgroups_open_current(void);
 void cgroups_set_memory_limit(int64_t bytes);
 
 #endif
